daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/NodeInfo.scala

```diff
@@ -61,16 +63,26 @@ import org.apache.daffodil.xml.XMLUtils
  * can answer questions about how two types are related. It can find the
  * least general supertype, or most general subtype of two types.
  */
-sealed abstract class TypeNode(parentsArg: => Seq[NodeInfo.Kind], childrenArg: => Seq[NodeInfo.Kind])
-  extends Serializable
+sealed abstract class TypeNode private (
+  val parents:  Seq[NodeInfo.Kind],
+  @TransientParam childrenArg: Delay[Seq[NodeInfo.Kind]])
+  extends PreSerialization
   with NodeInfo.Kind {
 
-  def this(parentArg: NodeInfo.Kind, childrenArg: => Seq[NodeInfo.Kind]) = this(Seq(parentArg), childrenArg)
-  def this(parentArg: NodeInfo.Kind) = this(Seq(parentArg), Seq(NodeInfo.Nothing))
+  def this(parents: Seq[NodeInfo.Kind], children: => Seq[NodeInfo.Kind]) = this(parents, Delay(children))
+  def this(parentArg: NodeInfo.Kind, childrenArg: => Seq[NodeInfo.Kind]) = this(Seq(parentArg), Delay(childrenArg))
+  def this(parentArg: NodeInfo.Kind) = this(Seq(parentArg), Delay(Seq[NodeInfo.Kind](NodeInfo.Nothing)))
 
   // Eliminated a var here. Doing functional graph construction now below.
-  final override lazy val parents = parentsArg
-  final override lazy val children = childrenArg
+  final override lazy val children = childrenArg.value
+
+  override def preSerialization = {
+    super.preSerialization
+    children
```

Comment:
We still require this preSerialization stuff to force the evaluation of these Delay values so that things can be garbage collected. Which means that if a user doesn't serialize the parser then things can't garbage collected. Or at least, they can't be garbage collected until a parse/unparse is run and these Delay's are used and evaluated. Which raises the issue, what if there are some Delay's that just happen to not be used? A single unevaluated Delay could hold on to a compile time object, which because it's a tree could hold onto basically the entire compile time stuff.

I think we have another bug for forcing all these serializtion things without actually serializing, so that maybe doesn't necessarily need to be done as part of this bug. But it's possible we won't see a difference until that's fixed.

Note if we didn't rely on writeObject to call preSerialization and we called it ourselves, I think the writeObject stuff could all go away, so fixing this would also simplify that I think.

Reply:
I am adding a counter to Delay objects so we know how many there are, and after serialization an assertion checks that there the invariant that there are zero unevaluated ones.